### ISSUE TITLE
Allow `Range` trait to accept arbitrary `__index__`-aware objects.

### ISSUE THREAD

**Issue by mdickinson** [enthought/traits#292]
Unlike the `Int` trait, the `Range` trait (for the integer range case) won't accept arbitrary objects with an `__index__` method. In particular, this affects some NumPy integer types. It would be useful to modify the `Range` trait to be more permissive.

```
>>> import numpy as np
>>> from traits.api import *
>>> class A(HasStrictTraits):
...     foo = Int
...     bar = Range(-128, 127)
... 
>>> a = A()
>>> x = np.int16(53)
>>> a.foo = x
>>> a.bar = x
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "build/bdist.macosx-10.6-x86_64/egg/traits/trait_handlers.py", line 172, in error
traits.trait_errors.TraitError: The 'bar' trait of an A instance must be -128 <= an integer <= 127, but a value of 53 <type 'numpy.int16'> was specified.
>>> x.__index__()
53
```

**Comment by mdickinson** [comment]
A similar issue exists for floats:
```
>>> class B(HasStrictTraits):
...     foo = Float
...     bar = Range(0.0, 1.0)
... 
>>> x = np.float32(0.43)
>>> b = B()
>>> b.foo = x
>>> b.bar = x
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "build/bdist.macosx-10.6-x86_64/egg/traits/trait_handlers.py", line 172, in error
traits.trait_errors.TraitError: The 'bar' trait of a B instance must be 0.0 <= a floating point number <= 1.0, but a value of 0.43000001 <type 'numpy.float32'> was specified.
```

**Comment by mdickinson** [comment]
This was fixed in #582, and the fix was released in Traits 6.0.
